Added 'angle' option to Polar Charts

slinhart opened the pull request:
I came across this inactive [Issue](https://github.com/chartjs/Chart.js/pull/4753) and picked it up.

The image comparison test I added is not passing even though the diff is 0% 🤷‍♂️  
Also, I imagine this will need a documentation update. Should the "Scriptable" column be added to the options table for Polar Charts? The first sentence for Polar Charts might also need updated since it reads

> Polar area charts are similar to pie charts, but each segment has the same angle - the radius of the segment differs depending on the value.

etimberg commented:
I looked at the test failure quickly. Looks like 1px differs against the original image. Not sure what the different pixel is since it didn't really stand out on the diff image. I noticed that labels on the index were still present. Perhaps removing them as well will remove the single pixel diff.

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -211,12 +223,29 @@ module.exports = function(Chart) {
+			var defaultAngle = (2 * Math.PI) / count;
+			var angleOption = me.chart.options.angle || defaultAngle;
+			var resolve = helpers.options.resolve;
+			var dataset = me.getDataset();
```
I would change the method signature to only pass `index` as argument and read the value here:

```javascript
_computeAngle: function(index) {
   //...

  var dataset = me.getDataset();
  var value = dataset.data[index];

  //...
}
```

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -133,6 +138,14 @@ module.exports = function(Chart) {
 			meta.count = me.countVisibleElements();
 
+			var start = opts.startAngle || 0;
+			for (var i = 0; i < meta.count; i++) {
```
Are you sure we only want to iterate until `meta.count` representing the number of visible elements, which are not necessary the first elements?

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -211,12 +223,29 @@ module.exports = function(Chart) {
 			var count = this.getMeta().count;
-			if (count > 0 && !isNaN(value)) {
-				return (2 * Math.PI) / count;
+			if (count <= 0 || isNaN(value)) {
```
I think we also need to return `0` is the element is hidden

slinhart commented:
@simonbrunel Thank you very much for the detailed review. I've updated the code based on your feedback. As far as the tests, I'm still having trouble getting those to pass. I've removed all the text from the chart and the image comparison is still failing (displays 0px diff in debug view). I added another test for 'undefined', but for the 'scriptiable' test I was unsure how to define this in the JSON file.

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -123,7 +123,12 @@ module.exports = function(Chart) {
 			var minSize = Math.min(chartArea.right - chartArea.left, chartArea.bottom - chartArea.top);
+			var start = opts.startAngle || 0;
+			var starts = [];
+			var angles = [];
```
I changed my previous comment a bit too late: 

```
var starts = me._starts = [];
var angles = me._angles = [];
```

And lines 148-149 can be removed :)

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -211,12 +225,30 @@ module.exports = function(Chart) {
+			var value = dataset.data[index];
+			var isHidden = meta.data[index].hidden;
+
+			if (count <= 0 || isNaN(value) || isHidden) {
```
When referenced only one time, I think it's preferable to avoid local variable for short assignment. Also, I don't think we need to check `count <= 0` anymore because (if I'm not wrong) if `count <= 0` then the value is either `NaN` or `hidden`.

```javascript
if (isNaN(dataset.data[index]) || meta.data[index].hidden) {
  return 0;
}
```

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -211,12 +225,30 @@ module.exports = function(Chart) {
 		},
 
-		calculateCircumference: function(value) {
+		_computeAngle: function(index) {
```
We started to be even more explicit about private stuff by adding the following comment above:

```javascript
/**
 * @private
 */
_computeAngle: function(index) {
```

simonbrunel commented on test/fixtures/controller.polarArea/angle-array.json:
```diff
@@ -0,0 +1,32 @@
+{
+    "debug": true,
```
Can be removed, it's only used for debugging and is likely the reason why unit tests are failing on Travis.

simonbrunel commented on test/fixtures/controller.polarArea/angle-array.json:
```diff
@@ -0,0 +1,32 @@
+{
+    "debug": true,
+    "config": {
+        "debug": true,
```
This one can be removed because it doesn't do anything

simonbrunel commented on src/controllers/controller.polarArea.js:
```diff
@@ -164,8 +175,8 @@ module.exports = function(Chart) {
 			// var negHalfPI = -0.5 * Math.PI;
 			var datasetStartAngle = opts.startAngle;
 			var distance = arc.hidden ? 0 : scale.getDistanceFromCenterForValue(dataset.data[index]);
-			var startAngle = datasetStartAngle + (circumference * visibleCount);
```
Travis complains that `visibleCount` is not anymore used and it seems that the whole block from line 167 to 173 can be removed. Does that make sense?

slinhart replied:
Oh right, thanks "Travis" 😅

etimberg commented:
@simonbrunel to answer your questions:

1. What happens when multiple datasets exist in the polar area chart? I think we only support 1 making this unnecessary. If multiple datasets work, then we should support this.
2. I don't think that's necessary for the same reasons as above. 
3. Yes
4. I would say radians to match other APIs in the polar area chart. We can change to degrees in v3

simonbrunel commented:
@etimberg I think this PR is almost ready to be merged, the only reason it's not is the choice of the option, `chart.options.angle` at the root, which could be confusing with other concepts. Should we prefer `chart.options.element.arc.angle` since this feature impact the arc angle?

etimberg commented:
@simonbrunel I think `chart.options.element.arc.angle` is fine since it keeps it away from other things.

slinhart commented:
@simonbrunel Sure, I'll update that
